beacon_node/beacon_chain/src/fetch_blobs.rs

```diff
@@ -125,136 +164,160 @@ pub async fn fetch_and_process_engine_blobs<T: BeaconChainTypes>(
+                let BlobAndProofV2 { blob, proofs } = blob_and_proof;
+                (blob, proofs)
+            })
+        })
+        .unzip();
 
-        Some(data_columns_receiver)
+    let num_fetched_blobs = blobs.len();
+    metrics::observe(&metrics::BLOBS_FROM_EL_RECEIVED, num_fetched_blobs as f64);
+
+    if num_fetched_blobs == 0 {
+        debug!(num_expected_blobs, "No blobs fetched from the EL");
+        inc_counter(&metrics::BLOBS_FROM_EL_MISS_TOTAL);
+        return Ok(None);
     } else {
-        if !blobs_to_import_and_publish.is_empty() {
-            publish_fn(BlobsOrDataColumns::Blobs(blobs_to_import_and_publish));
-        }
+        inc_counter(&metrics::BLOBS_FROM_EL_HIT_TOTAL);
+    }
 
-        None
-    };
+    // Partial blobs response isn't useful for PeerDAS, so we don't bother building and publishing data columns.
+    if num_fetched_blobs != num_expected_blobs {
+        debug!(
+            info = "Unable to compute data columns",
+            num_fetched_blobs, num_expected_blobs, "Not all blobs fetched from the EL"
+        );
+        return Ok(None);
+    }
+
+    if chain
+        .canonical_head
+        .fork_choice_read_lock()
+        .contains_block(&block_root)
+    {
+        // Avoid computing columns if block has already been imported.
+        debug!(
+            info = "block has already been imported",
+            "Ignoring EL blobs response"
+        );
+        return Ok(None);
+    }
+
+    let custody_columns = spawn_compute_and_publish_data_columns_task(
+        &chain,
+        block.clone(),
+        blobs,
+        proofs,
+        custody_columns_indices,
+        publish_fn,
+    )
+    .await?;
 
     debug!(num_fetched_blobs, "Processing engine blobs");
 
     let availability_processing_status = chain
         .process_engine_blobs(
             block.slot(),
             block_root,
-            fixed_blob_sidecar_list.clone(),
-            data_columns_receiver_opt,
+            EngineGetBlobsOutput::DataColumns(custody_columns),
         )
         .await
         .map_err(FetchEngineBlobError::BlobProcessingError)?;
 
     Ok(Some(availability_processing_status))
 }
 
-/// Spawn a blocking task here for long computation tasks, so it doesn't block processing, and it
-/// allows blobs / data columns to propagate without waiting for processing.
-///
-/// An `mpsc::Sender` is then used to send the produced data columns to the `beacon_chain` for it
-/// to be persisted, **after** the block is made attestable.
-///
-/// The reason for doing this is to make the block available and attestable as soon as possible,
-/// while maintaining the invariant that block and data columns are persisted atomically.
-fn spawn_compute_and_publish_data_columns_task<T: BeaconChainTypes>(
+/// Offload the data column computation to a blocking task to avoid holding up the async runtime.
+async fn spawn_compute_and_publish_data_columns_task<T: BeaconChainTypes>(
```

Comment:
```suggestion
async fn compute_and_publish_data_columns_task<T: BeaconChainTypes>(
```
Name is misleading now

Reply:
It is still spawning a blocking task, but I guess it's not relevant to the caller, I'll update the method name.